backend/editor/controllers/search_controller.py

```diff
@@ -0,0 +1,374 @@
+            filter_search_terms.append(filter_search_term)
+        else:
+            name_search_terms.append(search_term)
+
+    return Query(project_id, name_search_terms, filter_search_terms)
+
+
+def build_lucene_name_search_query(search_value: str) -> str | None:
+    """
+    The name search term can trigger two types of searches:
+    - if the search value is in the format `language_code:raw_search_value`,
+      it triggers a search on the tags_ids_{language_code} index
+    - else it triggers a search on the tags_ids index
+
+    If the `raw_search_value` is surrounded by quotes, the search will be exact.
+    Otherwise, the search is fuzzy when the search value is longer than 4 characters
+    (the edit distance depends of the length of the search value)
+    """
+    language_code = None
+
+    if len(search_value) > 2 and search_value[2] == ":" and search_value[0:2].isalpha():
+        language_code, search_value = search_value.split(":", maxsplit=1)
+        language_code = language_code.lower()
+
+    def get_search_query() -> str | None:
+        if search_value.startswith('"') and search_value.endswith('"'):
+            return search_value if len(search_value) > 2 else None
+
+        if language_code is not None:
+            normalized_text = parser_utils.normalize_text(search_value, language_code)
+        else:
+            normalized_text = parser_utils.normalize_text(search_value)
+
+        # If normalized text is empty, no searches are found
+        if normalized_text.strip() == "":
+            return None
+
+        tokens = normalized_text.split("-")
+
+        def get_token_query(token: str) -> str:
+            token = "+" + token
+            if len(token) > 10:
+                return token + "~2"
+            elif len(token) > 4:
+                return token + "~1"
+            else:
+                return token
+
+        return "(" + " ".join(map(get_token_query, tokens)) + ")"
+
+    search_query = get_search_query()
+
+    if search_query is None:
+        return None
+
+    if language_code is not None:
+        search_query = f"tags_ids_{language_code}:{search_query}"
+
+    return search_query
+
+
+def build_cypher_name_search_term(project_id: str) -> tuple[str, str, str]:
+    query = f"""
+            CALL db.index.fulltext.queryNodes("{project_id}_SearchTagsIds", $search_query)
+            YIELD node, score
+            WHERE score > 0.1
+            WITH node.id AS nodeId
+            WITH COLLECT(nodeId) AS nodeIds
+        """
+
+    where_clause = "n.id IN nodeIds"
+
+    order_clause = "WITH n, apoc.coll.indexOf(nodeIds, n.id) AS index ORDER BY index"
+
+    return query, where_clause, order_clause
+
+
+def build_cypher_query(
+    query: Query, skip: int, limit: int
+) -> tuple[str, str, dict[str, str]] | None:
+    lucene_name_search_queries = list(
+        filter(
+            lambda q: q is not None, map(build_lucene_name_search_query, query.name_search_terms)
+        )
+    )
```

Comment:
I prefer to add comment whenever code is a bit compact:
```suggestion
    # build part of the query doing full text searcf
    lucene_name_search_queries = list(
        filter(
            lambda q: q is not None, map(build_lucene_name_search_query, query.name_search_terms)
        )
    )
```